kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaChanges.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (2025) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.util;
+
+import io.delta.kernel.types.StructField;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * SchemaChanges encapsulates a list of added, removed, renamed, or updated fields in a schema
+ * change. Updated fields include renamed fields, moved fields, type changes, nullability changes,
+ * and metadata attribute changes.
+ *
+ * <p>ToDo: Possibly track moves/renames independently
+ */
+class SchemaChanges {
+  private List<StructField> addedFields;
+  private List<StructField> removedFields;
+  private List<Tuple2<StructField, StructField>> updatedFields;
+
+  private SchemaChanges(
+      List<StructField> addedFields,
+      List<StructField> removedFields,
+      List<Tuple2<StructField, StructField>> updatedFields) {
+    this.addedFields = addedFields;
+    this.removedFields = removedFields;
+    this.updatedFields = updatedFields;
+  }
+
+  static class Builder {
+    private List<StructField> addedFields = new ArrayList<>();
+    private List<StructField> removedFields = new ArrayList<>();
+    private List<Tuple2<StructField, StructField>> updatedFields = new ArrayList<>();
+
+    public Builder withAddedField(StructField addedField) {
+      addedFields.add(addedField);
+      return this;
+    }
+
+    public Builder withRemovedField(StructField removedField) {
+      removedFields.add(removedField);
+      return this;
+    }
+
+    public Builder withUpdatedField(StructField existingField, StructField newField) {
+      updatedFields.add(new Tuple2<>(existingField, newField));
+      return this;
+    }
+
+    public SchemaChanges build() {
+      return new SchemaChanges(addedFields, removedFields, updatedFields);
+    }
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  /* Added Fields */
+  public List<StructField> addedFields() {
```

Comment:
I could be wrong, but I think associating each of these with the column path of where it is helps with lot of things: like identifying renames, relocation across nested structs etc.

Reply:
It's possible, I'm still thinking that SchemaChanges may become aware of changes based on ID. The reason is that we know some evolutions inherently cannot happen safely without IDs (such as renames, moves, drops). So it would not be overly restrictive to force SchemaChanges to just have that baked in?